File: src/data-structures/linear/linked-lists/py/linked_list.py
from __future__ import annotations


class Node:
    next: Node = None
    value: any = None

    def __init__(self, value: any = None, next: Node = None) -> None:
        self.value = value
        self.next = next


class LinkedList:
    """
    Operations
    Traversal - access each element of the linked list
    Insertion - adds a new element to the linked list
    Deletion - removes the existing elements
    Search - find a node in the linked list
    Sort - sort the nodes of the linked list
    """

    head: Node = None
    value: any = None
    size: int = 0

    def __init__(self, value: any = None, next: Node = None):
        self.next = next
        self.value = value
        self.head = Node()
        self.size = 0

    def get_size(self) -> int:
        return self.size

    def prepend(self, value: any):
        """
        adds an item the beginning of the list
        """
        next = self.head.next
        node = Node(value, next)
        self.head.next = node
        self.size += 1

    def insert_bulk(self, values: list[value]):
        for v in values:
            self.insert(v)

    def insert(self, value: any):
        self._insert(Node(value))

    def _insert(self, node: Node):
        """
        adds an item to end of the list
        """
        current = self.head

        if current.next is None:
            current.next = node
        else:
            while current.next != None:
                current = current.next

            current.next = node
        self.size += 1

    def search(self, value: any) -> bool:
        """
        searches a linked list to check if a value exists
        """
        current = self.head

        while current is not None:
            if current.value == value:
                return True
            else:
                current = current.next

        return False

    def get_by_index(self, index: int) -> Node:
        """
        gets an item at the specified index
        """
        if index > self.size or index < 0:
            raise IndexError("Index out of bounds of list size")

        current = self.head.next
        count = 0
        while current:
            if count == index:
                return current
            count += 1
            current = current.next

        return None

    def insert_at(self, value: any, index: int) -> Node:
        if index < 0 or index > self.size:
            raise IndexError("Index out of bounds of List")

        current = self.head
        for i in range(index):
            current = current.next

        next = current.next
        node = Node(value=value, next=next)
        current.next = node

        return node

    def __iter__(self):

        return self

    def __next__(self):
        node = self.head

        if node:
            return node

File: src/data-structures/linear/linked-lists/js/LinkedList.js
//node class
class Node {
  constructor(value, next) {
    this.value = value;
    this.next = null;
  }
}

//ll class
class LinkedList {
  constructor() {
    //initialize the head and the length properties
    this.head = null;
    this.tail = null;
    this.size = 0;
  }

  //append (end)  O(1) adds a new element to the end of the list
  append(value) {
    if (!this.head) {
      this.prepend(value);
    } else {
      //create a new node object with the element and next properties
      const node = new Node(value);
      this.tail.next = node;
      this.tail = node;
      this.size++;
    }
    return this;
  }

  //Prepend (beginning)  O(1) adds to the beginning of the list
  prepend(value) {
    const newNode = new Node(value);

    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
      this.size++; //1
    } else {
      newNode.next = this.head;
      this.head = newNode;
      this.size++;
    }
  }

  //Print List (value) O(n)
  getByIndex(index) {
    // if the index is out of bounds, return null
    if (index < 0 || index >= this.size) return null;
    let current = this.head;
    for (let i = 0; i < index; i++) {
      current = current.next;
    }
    return current;
  }

  //print all elements of the linked list
  printList() {
    const result = [];
    let current = this.head;
    let counter = 0;
    while (counter < this.size) {
      current = current.next;
      counter++;
    }
    return result;
  }

  //inserts a new element at a given index
  insert(value, index) {
    if (index < 0 || index > this.size) {
      return -1;
    } else if (index === 0) {
      this.prepend(value);
    } else if (data === this.size) {
      this.append(value);
    } else {
      const newNode = new Node(value);
      let current = this.head;
      let counter = 0;

      while (counter < index - 1) {
        counter++;
        current = current.next;
      }
      let temp = current.next;
      current.next = newNode;
      newNode.next = temp;
      this.size++;
    }
    return this;
  }

  //removes the last element of the list
  delete(index) {
    // if the head is null, return null
    if (index < 0 || index >= this.size) {
      return -1;
    } else if (index === 0) {
      this.removeHead();
    } else if (index === this.size - 1) {
      this.removeTail();
    } else {
      let current = this.head;
      let count = 0;
      while (count < index - 1) {
        current = current.next;
        count++;
      }
      current.next = current.next.next;
      this.size--;
    }
  }

  // reverse
  reverse(head) {
    let currentNode = this.head;
    let prevNode = null;
    let nextNode = null;

    while (currentNode) {
      // While currentNode is not equal to null

      // Stash next node.
      nextNode = currentNode.next;

      // Update next node of the current node to link to previous node.
      currentNode.next = prevNode;

      // Move prevNode and currNode nodes one step forward.
      prevNode = currentNode;
      currentNode = nextNode;
    }

    // Reset head and tail.
    this.tail = this.head;
    this.head = prevNode;

    return this;
  }
}

LinkedList.fromValues = (...values) => {
  const myLinkedList = new LinkedList();
  for (let i = values.length - 1; i >= 0; i--) {
    myLinkedList.prepend(values[i]);
  }
  return myLinkedList;
};

module.exports = LinkedList;
